### PR TITLE
[FlightRPC][Java] Add token authentication to Flight JDBC

### PR DESCRIPTION
**Pull request by jcralmeida**

### CODE REVIEW

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightServerTestRule.java [hunk 1]
```diff
@@ -170,18 +134,88 @@ private FlightServer getStartServer(Function<Location, FlightServer> newServerFr
...
+      this.authentication = authentication;
+      return this;
+    }
+
+    /**
+     * Builds the {@link FlightServerTestRule} using the provided values.
+     * @return  a {@link FlightServerTestRule}.
+     */
+    public FlightServerTestRule build() {
+      if (authentication instanceof UserPasswordAuthentication) {
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightServerTestRule.java:208]: Can we refactor it so we don't need to use `instanceof`? Maybe add a getProperties() to the Authentication interface

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ArrowDatabaseMetadataTest.java [hunk 2]
```diff
@@ -269,6 +269,18 @@
   private static Connection connection;
 
   static {
+    UserPasswordAuthentication authentication =
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ArrowDatabaseMetadataTest.java:272]: I think we can keep a `FlightServerTestRule.createNewTestRule` like before that uses the builder internally.

Most test files changed to this pattern of building the FlightServerTestRule manually, but we can leave them unmodified

**Review by rafael-telles — CHANGES_REQUESTED** [review]

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightServerTestRule.java (hunk 1 above)
Comment (jcralmeida) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightServerTestRule.java]: Fixed

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ArrowDatabaseMetadataTest.java (hunk 2 above)
Comment (jcralmeida) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ArrowDatabaseMetadataTest.java]: Fixed

**Review by rafael-telles — APPROVED** [review]

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/TokenAuthenticationTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        .authentication(new TokenAuthentication.Builder()
+            .token("1234")
+            .build())
+        .producer(FLIGHT_SQL_PRODUCER)
+        .build();
+  }
+
+  @Test(expected = SQLException.class)
+  public void connectUsingTokenAuthenticationShouldFail() throws SQLException {
+    Connection connection = FLIGHT_SERVER_TEST_RULE.getConnectionFromToken("invalid");
```
Comment (jduo) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/TokenAuthenticationTest.java:48]: try with resources here and in the test below please

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/TokenAuthenticationTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    Assert.fail();
+  }
+
+  @Test
+  public void connectUsingTokenAuthenticationShouldSuccess() throws SQLException {
+    Connection connection = FLIGHT_SERVER_TEST_RULE.getConnectionFromToken("1234");
+    Assert.assertFalse(connection.isClosed());
```
Comment (jduo) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/TokenAuthenticationTest.java:55]: I'd think an exception would have been thrown if the connection could not be created.

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/TokenAuthenticationTest.java (hunk 3 above)
Comment (jcralmeida) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/TokenAuthenticationTest.java]: Fixed

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/TokenAuthenticationTest.java (hunk 4 above)
Comment (jcralmeida) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/TokenAuthenticationTest.java]: But this test case is for a valid connection. The one above is testing the scenario you mentioned

**Review by mcraealex-bq — COMMENTED** [review]
Beyond the unit tests, how is the authentication being tested? Are you guys in a position to hit an endpoint yet?

**Review by jduo — COMMENTED** [review]
Looks good to me.

**Comment by jcralmeida** [comment]
We only have the unit test @mcraealex-bq . I don't think we hit that point yet

**Pull request merged**